azkaban-web-server/src/main/java/azkaban/imagemgmt/services/ImageTypeServiceImpl.java

```diff
@@ -45,10 +45,10 @@ public ImageTypeServiceImpl(final ImageTypeDao imageTypeDao,
   }
 
   @Override
-  public int createImageType(ImageMetadataRequest imageMetadataRequest) throws IOException,
+  public int createImageType(final ImageMetadataRequest imageMetadataRequest) throws IOException,
       ImageMgmtException {
     // Convert input json payload to image type object
-    ImageType imageType = converterUtils
+    final ImageType imageType = this.converterUtils
         .convertToModel(imageMetadataRequest.getJsonPayload(), ImageType.class);
```

Comment:
I don't think we should convert json into model classes as there can be internal fields in the model that should not be exposed to users.
Please, take a look at the `Space` endpoints in the experimental APIs as an example of the approach we're following over there. We're also using DTOs but a bit different, in a nutshell, for data going from Servlets to Services we're using DTOs annotated with java bean validation annotations for easy validation on the Service. From Services to Servlets we're also using response DTOs: the DAO layer is currently returning model objects and those get converted into DTOs to return to Servlets.

I can think of the following advantages for this approach:
-easy to understand the request parameters the user need to provide (just by looking at the request DTO fields)
-faster to create new endpoints
-easy to communicate specific error messages (let the user know exactly what's the wrong parameter as opposed to give a general message).
-by using request/response DTOs it's not possible for unwanted/private fields to get parsed from requests or be included in responses. 
-the service is decoupled from the specific data format (json, xml, etc) users use to send the data.

Reply:
Agreed. All these are valuable points. The best way is to have API specific data model and model on top of database so that both can evolve independently. This would require to and fro conversion from API model(DTO) to database model with the need of additional conversions. I thought about it during the implementation, but considering the scope of MPV did not feel the need to such structure as the image management functionality is completely within the scope of Azkaban and not driven by any user specific scope. So there are very less possibility of mismatch between API input and the corresponding data model. However, we can definitely iterate on top of it in future considering the timeline and scope of MVP.  In the current codebase here are the things followed. 
1. json to database model conversion is one way. Did not introduce API specific model here as the input json and the database model is having exactly one to one mapping. There are very minute differences in certain cases, there API specific request DTO is used and separate model for database.
2. While sending response from the database same database model is returned because response and database model is having exactly same correspondence.
3. Input json is converted to model in the service layer and javax validation is applied on top of that. As mentioned earlier in certain cases json gets converted to request DTO (separate from database model) and javax validation annotation is applied there. This is done just to take additional API input. Also with this structure we can write complete business logic in the service layer such as payload to model conversion, validation and this also helps writing unit test for service layer. 
4. Image management APIs are very much Azkaban specific and not driven by any user usecase and 95% of the time there is exact correspondence between API input and database model. Hence, this structure is followed for the MVP.

Reply:
I can give you an example where a user can tamper with the data using current approach of parsing json request with model classes as opposed to DTOs.
When updating a ramp up plan a user can provide an id parameter in the json object. Because we're using the model class for parsing this will mean that a user is able to modify any ramp up plan where they don't have permissions.

We have been working on the experimental APIs for a few months now, we have faced most of the problems reflected in the comments of this PR(e.g: validation, error handling) and there have been a few iterations on the implementation to resolve them. Please let's leverage that experience and improve the approaches already followed there rather than start all over again. 

By following the patterns being used in the experimental APIs we're speeding up the process of adding these new APIs considerably, so in any case its adoption will help to meet MVP deadlines not the opposite. 
Also, MVP should not be a reason for delaying adoption of those patterns because containerized executions is not an experimental feature, this is definitely something that will be deployed to prod and merge into master branch right after it's finished so the more code we add with current approach the more code we'll need to refactor very soon.